### PR TITLE
Tracking annotations for (dis)appearances

### PR DESCRIPTION
**Pull request by ilastikdev**
requires hytra PR:
https://github.com/chaubold/hytra/pull/29

GUI allows the user to input (dis)appearance labels for objects of tracks:
only APPEARANCE labels for FIRST objects of  tracks and
only DISAPPEARANCE labels for LAST objects of tracks will be used
( (dis)appearance labels for intermediate objects of tracks will be ignored).

### CODE REVIEW

**Comment by ilastikdev** [comment]
This will be done by new versions of projects used by existing tests.

**Comment by ilastikdev** [comment]
last commit adds a project file for the required test of appearances/disappearances

**Comment by chaubold** [comment]
I assume old structured learning projects can still be loaded, but as they do not contain appearance/disappearance annotations ilastik will understand any annotation as "user started labeling in the middle of the track", right?

File: ilastik/applets/tracking/structured/opStructuredTracking.py [hunk 1]
```diff
@@ -264,6 +287,18 @@ def _runStructuredLearning(self,
                                             # print "[structuredTrackingGui] Increasing max nearest neighbors! LABELS/MERGERS t:{} id:{}".format(time-1, int(previous_label))
                                             break
 
+                                    if type[0] in ["FIRST", "SINGLETON(FIRST_LAST)"] and time in self.appearances.keys() and label in self.appearances[time].keys() and track in self.appearances[time][label].keys() and self.appearances[time][label][track]:
+                                        # print("---> appearance",time,label,track)
+                                        if (time, int(label)) in list(hypothesesGraph._graph.node.keys()):
+                                            hypothesesGraph._graph.node[(time, int(label))]['appearance'] = True
+                                            logger.info("[structuredTrackingGui] APPEARANCE: {} {}".format(time, int(label)))
```
Comment (chaubold) [inline ilastik/applets/tracking/structured/opStructuredTracking.py]: Please downgrade this to `logger.debug` as it is spamming the console and shouldn't be seen by users.

**Comment by ilastikdev** [comment]
all tests execute
needs
https://github.com/chaubold/hytra/pull/32

**Comment by ilastikdev** [comment]
Gui mechanism for (dis)appearances was discussed with Anna,
also Fred did not like anymore new buttons,
so we decided on drop down menu on right-click.

An object is considered a non-(dis)appearance unless the user annotates it - and the user is only allowed to do so if the object is first/last on a track.

**Comment by chaubold** [comment]
If you decrease the log level as noted above then I'll merge it

**Comment by chaubold** [comment]
[structured13.ilp.zip](https://github.com/ilastik/ilastik/files/1488044/structured13.ilp.zip)

Hmm. In the attached project I've purposely labeled some tracks to start/stop in the middle of the dataset to experiment with the new buttons. Why can't I annotate the blue object in frame 1 as a disappearance? That should be possible, as it is not part of the merger in the next frame...

**Comment by ilastikdev** [comment]
Only the last object of a tack can be labelled as a disappearance.

**Comment by chaubold** [comment]
it is the last object of the track

**Comment by ilastikdev** [comment]
This is a threading issue: annotating helper runSubtracking is run in a separate thread and should be followed by saveAnnotations.

However, runSubtracking sometimes executes one of the last addObjectToTrack calls after saveAnnotations has been called.

(code runs in correct sequence with LAZYFLOW_THREADS=0 :)

**Pull request merged**